runtime/bcutil/ROMClassCreationContext.hpp

```diff
@@ -384,11 +399,38 @@ class ROMClassCreationContext
 	{
 		if (!isRedefining() && !isRetransforming()) {
 			if (NULL != _className) {
+				U_16 classNameLenToCompare0 = _classNameLength;
+				U_16 classNameLenToCompare1 = classNameLength;
+				BOOLEAN misMatch = FALSE;
 				if (isClassHidden()) {
-					/* for hidden class className has ROM address appended at the end, _className does not have that */
-					classNameLength = (U_16)_classNameLength;
+					if (isROMClassShareable()) {
+						U_8* lambdaClass0 = (U_8*)getLastDollarSignOfLambdaClassName((const char*)_className, _classNameLength);
+						U_8* lambdaClass1 = (U_8*)getLastDollarSignOfLambdaClassName((const char*)className, classNameLength);
+						if ((NULL != lambdaClass0) 
+							&& (NULL != lambdaClass1)
+						) {
+							/**
+							 * Lambda class has class name: HostClassName$$Lambda$<IndexNumber>/0x0000000000000000. 
+							 * Do not need to compare the IndexNumber as it can be different from run to run.
+							 */
+							classNameLenToCompare0 = (U_16)(lambdaClass0 - _className + 1);
+							classNameLenToCompare1 = (U_16)(lambdaClass1 - className + 1);
+						} else if ((NULL == lambdaClass0)
+								&& (NULL == lambdaClass1)
+						) {
+							/* for hidden class className has ROM address appended at the end, _className does not have that */
+							classNameLenToCompare1 = (U_16)_classNameLength;
+						} else {
+							misMatch = TRUE;
+						}
+					} else {
+						/* for hidden class className has ROM address appended at the end, _className does not have that */
+						classNameLenToCompare1 = (U_16)_classNameLength;
+					}
 				}
-				if ((0 == J9UTF8_DATA_EQUALS(_className, _classNameLength, className, classNameLength))) {
+				if (misMatch || 
+					(0 == J9UTF8_DATA_EQUALS(_className, classNameLenToCompare0, className, classNameLenToCompare1))
```

Comment:
J9UTF8_DATA_EQUALS returns a boolean, pls use `!` instead of `0 ==`

Reply:
Fixed.